awslabs/soci-snapshotter
# Add Referrers API support

This commit introduces ORAS Referrers API support to the snapshotter. `--soci-index-digest` is no longer necessary as the snapshotter will invoke the Referrers API.

Signed-off-by: Rishabh Singhvi <rdpsin@amazon.com>

*Issue #, if available:*
Resolves #44 

*Description of changes:*

*Testing performed:*

`make && make check && make test && make integration`


1) Start the snapshotter
```shell
$ sudo ./soci-snapshotter-grpc &> logs &
```

2) `rpull` the image, but don't provide `--soci-index-digest`

```shell
$ sudo ./soci image rpull --user "$REGISTRY_USER:$REGISTRY_PASS" $IMAGE
fetching sha256:3fd97d24... application/vnd.docker.distribution.manifest.v2+json
fetching sha256:aec97901... application/vnd.docker.container.image.v1+json
```

3) Check `mount`

```shell
$ mount 
/local/home/rdpsin/soci/out/soci on /var/lib/soci-snapshotter-grpc/snapshotter/snapshots/1/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
/local/home/rdpsin/soci/out/soci on /var/lib/soci-snapshotter-grpc/snapshotter/snapshots/2/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
/local/home/rdpsin/soci/out/soci on /var/lib/soci-snapshotter-grpc/snapshotter/snapshots/3/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
...
```

4) Check the logs to make sure the referrers API is being called

```
{"key":"default/1/extract-48576544-22Re sha256:08249ce7456a1c0613eafe868aed936a284ed9f1d6144f7d2d08c514974a2af9","level":"info","mountpoint":"/var/lib/soci-snapshotter-grpc/snapshotter/snapshots/1/fs","msg":"index digest not provided, making a Referrers API call to fetch list of indices","parent":"","time":"2022-09-15T14:22:01.056566326Z"}
{"digest":"sha256:2acf14636fb5afd8f862b7dff02fbbd053eead67b9aa0baf69e9ca5c233a2115","key":"default/1/extract-48576544-22Re sha256:08249ce7456a1c0613eafe868aed936a284ed9f1d6144f7d2d08c514974a2af9","level":"info","mountpoint":"/var/lib/soci-snapshotter-grpc/snapshotter/snapshots/1/fs","msg":"fetching SOCI artifacts using index descriptor","parent":"","time":"2022-09-15T14:22:01.674535703Z"}
```
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
